src/web/page/pages.rs

```diff
@@ -0,0 +1,64 @@
+use super::TEMPLATE_DATA;
+use crate::docbuilder::Limits;
+use iron::{headers::ContentType, response::Response, status::Status, IronResult};
+use serde::Serialize;
+use tera::Context;
+
+/// The central trait that rendering pages revolves around, it handles selecting and rendering the template
+pub trait WebPage: Serialize + Sized {
+    /// Turn the current instance into a `Response`, ready to be served
+    // TODO: The potential for caching similar pages is here due to render taking `&Context`
+    fn into_response(self) -> IronResult<Response> {
+        let ctx = Context::from_serialize(&self).unwrap();
+        let rendered = dbg!(TEMPLATE_DATA.templates.load())
+            .render(Self::template(), &ctx)
+            .unwrap();
+
+        let mut response = Response::with((self.get_status(), rendered));
+        response.headers.set(Self::content_type());
+
+        Ok(response)
+    }
+
+    /// Get the name of the template to be rendered
+    fn template() -> &'static str;
+
+    /// Gets the status of the request, defaults to `Ok`
+    fn get_status(&self) -> Status {
+        Status::Ok
+    }
+
+    /// The content type that the template should be served with, defaults to html
+    fn content_type() -> ContentType {
+        ContentType::html()
+    }
+}
+
+/// The sitemap, corresponds with `templates/sitemap.xml`
+#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
+pub(crate) struct SitemapXml {
+    /// The releases to be displayed on the sitemap
+    pub releases: Vec<(String, String)>,
+}
+
+impl WebPage for SitemapXml {
+    fn template() -> &'static str {
+        "docsrs/sitemap.xml"
```

Comment:
It selects the template to be rendered

Reply:
Right, but why is it needed? Where is it called?

Reply:
It's called as part of the `.into_response()` function to get the template that the given struct needs in order to be correctly rendered

https://github.com/rust-lang/docs.rs/blob/2d0156b52daa74ec4daba132a78cb6736f82ae1e/src/web/page/pages.rs#L14

Reply:
Does it ever need to be dynamic? Could it be `const TEMPLATE: &'static str;` instead? (Given that it's an `&'static str` I assume it's not going to be dynamic, since that would require leaking). (I would say the same about the content type, but given it uses `vec![]` I assume it has no `const` constructors).

Reply:
It could be an associated static of some sort, but this does leave the flexibility for us to do whatever we want in the future

Reply:
> this does leave the flexibility for us to do whatever we want in the future

It's not flexible enough to be useful because it returns a `&'static str` (as mentioned by Nemo), and I don't think there's any point in trying to generalize the functionality without any idea of what changes we'd make. So I agree it should be an associated constant.